processing/src/main/java/org/apache/druid/segment/virtual/NestedFieldVirtualColumn.java

```diff
@@ -601,15 +606,76 @@ public VectorValueSelector makeVectorValueSelector(
     if (!(theColumn instanceof NestedDataComplexColumn)) {
 
       if (parts.isEmpty()) {
-        ColumnCapabilities capabilities = holder.getCapabilities();
         if (theColumn instanceof DictionaryEncodedColumn) {
-          return ExpressionVectorSelectors.castObjectSelectorToNumeric(
-              offset,
-              this.columnName,
-              theColumn.makeVectorObjectSelector(offset),
-              capabilities.toColumnType(),
-              expectedType
-          );
+          final VectorObjectSelector delegate = theColumn.makeVectorObjectSelector(offset);
+          return new VectorValueSelector()
+          {
+            private int currentOffsetId = ReadableVectorInspector.NULL_ID;
+            private long[] longs = new long[delegate.getMaxVectorSize()];
+            private double[] doubles = new double[delegate.getMaxVectorSize()];
+            private float[] floats = new float[delegate.getMaxVectorSize()];
+            private boolean[] nulls = new boolean[delegate.getMaxVectorSize()];
+
+            @Override
+            public long[] getLongVector()
+            {
+              computeNumbers();
+              return longs;
+            }
+
+            @Override
+            public float[] getFloatVector()
+            {
+              computeNumbers();
+              return floats;
+            }
+
+            @Override
+            public double[] getDoubleVector()
+            {
+              computeNumbers();
+              return doubles;
+            }
+
+            @Nullable
+            @Override
+            public boolean[] getNullVector()
+            {
+              computeNumbers();
+              return nulls;
+            }
+
+            @Override
+            public int getMaxVectorSize()
+            {
+              return delegate.getMaxVectorSize();
+            }
+
+            @Override
+            public int getCurrentVectorSize()
+            {
+              return delegate.getCurrentVectorSize();
+            }
+
+            private void computeNumbers()
+            {
+              if (currentOffsetId != offset.getId()) {
+                currentOffsetId = offset.getId();
+                final Object[] values = delegate.getObjectVector();
+                for (int i = 0; i < values.length; i++) {
+                  Number n = ExprEval.computeNumber(Evals.asString(values[i]));
+                  if (n != null) {
+                    longs[i] = n.longValue();
```

Comment:
It's not likely that all of these are going to be needed by the same query. Could the code be structured to avoid computing them all upfront?

Reply:
split up based on expected output type